src/backend/joanie/payment/models.py

```diff
@@ -1 +1,465 @@
+        default_currency=settings.DEFAULT_CURRENCY,
+        null=False,
+        blank=False,
+    )
+    recipient_name = models.CharField(
+        _("invoice recipient"), max_length=40, null=False, blank=False
+    )
+    recipient_address = models.TextField(
+        _("invoice address"), max_length=255, null=False, blank=False
+    )
+    localized_context = models.JSONField(
+        _("context"),
+        help_text=_("Localized data that needs to be frozen on invoice creation"),
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "joanie_invoice"
+        verbose_name = _("Invoice")
+        verbose_name_plural = _("Invoices")
+        constraints = [
+            models.CheckConstraint(
+                check=(
+                    models.Q(order__isnull=False)
+                    & models.Q(parent__isnull=True)
+                    & models.Q(total__gte=0)
+                )
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def state(self):
+        """
+        Process the state of the invoice
+        """
+        if self.balance.amount >= 0:
+            if (
+                self.invoiced_balance.amount == 0
+                and self.transactions_balance.amount == 0
+            ):
+                return payment_enums.INVOICE_STATE_REFUNDED
+
+            return payment_enums.INVOICE_STATE_PAID
+
+        return payment_enums.INVOICE_STATE_UNPAID
+
+    @property
+    def related_order(self):
+        """
+        Return the related order.
+        """
+        return self.order or self.parent.order
+
+    @property
+    def document(self):
+        """
+        Get the document related to the invoice instance;
+        """
+        # pylint: disable=no-member
+        document = InvoiceDocument(context_query=self.get_document_context())
+        return document.create(persist=False)
+
+    @property
+    def type(self):
+        """
+        Return the invoice type according to its total amount.
+        If total amount is positive, invoice type is "invoice"
+        otherwise it's "credit_note"
+        """
+        if self.total.amount < 0:  # pylint: disable=no-member
+            return payment_enums.INVOICE_TYPE_CREDIT_NOTE
+
+        return payment_enums.INVOICE_TYPE_INVOICE
+
+    @cached_property
+    def transactions_balance(self):
+        """
+        Process the transactions balance.
+
+        First we retrieve all transactions registered for the current invoice
+        and its children. Then we sum all transactions amount.
+        """
+        invoices = [self, *self.children.all()]
+        ids = [invoice.id for invoice in invoices]
+
+        amount = Transaction.objects.filter(invoice__id__in=ids).aggregate(
```

Comment:
I would say that this does the same in 1 query:
```suggestion
        amount = Transaction.objects.filter(Q(invoice__in=self.children.all()) | Q(invoice=self)).aggregate(
```